### ISSUE TITLE
"oscap-docker" doesn't work with default container storage on Fedora 26

### ISSUE THREAD

**Issue by jan-cerny** [OpenSCAP/openscap#794]
Fedora 26 has started to use overlay2 as a storage. Overlay has replaced devicemapper, which was used in older Fedoras. See https://fedoraproject.org/wiki/Changes/DockerOverlay2. 

Containers on devicemapper are mounted by `atomic.mount.DockerMount` into `[tmpdir]/rootfs`, but containers on overlay2 are mounted by `atomic.mount.DockerMount` into `[tmpdir]`, which isn't expected by oscap-docker.  It expects  `[tmpdir]/rootfs` always. See https://github.com/OpenSCAP/openscap/blob/maint-1.2/utils/oscap_docker_python/oscap_docker_util.py#L262 

As a result, OpenSCAP then tries to chroot to a non-existing directory, which leads to error of `oscap` utility, and subsequent tracebacks.

**Comment by jan-cerny** [comment]
For example, this command:

```
# oscap-docker image fedora xccdf eval --profile common /usr/share/xml/scap/ssg/content/ssg-fedora-xccdf.xml
```

Produces this output:
```
Command: oscap xccdf eval --profile common /usr/share/xml/scap/ssg/content/ssg-fedora-xccdf.xml failed!
Command returned exit code 1.
FAIL: 291:chdir: 2, No such file or directory
W: oscap: Can't receive message: 104, Connection reset by peer.
E: oscap: Recv: retry limit (0) reached.
OpenSCAP Error: Unable to receive a message from probe [oval_probe_ext.c:579]
Failed to create new OVAL agent session for: 'ssg-fedora-oval.xml'. [xccdf_session.c:919]

Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/docker/api/client.py", line 220, in _raise_for_status
    response.raise_for_status()
  File "/usr/lib/python3.6/site-packages/requests/models.py", line 909, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 404 Client Error: Not Found for url: http+docker://localunixsocket/v1.26/containers/tmpnxkmw9tj?v=False&link=False&force=False

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./oscap-docker", line 115, in <module>
    args.func()
  File "./oscap-docker", line 51, in scan
    result = OS.scan(self.args.scan_target, self.unknown_args)
  File "/home/jcerny/openscap/utils/oscap_docker_python/oscap_docker_util.py", line 265, in scan
    sys.stdout.write(self.helper._scan(chroot, scan_args))
  File "/home/jcerny/openscap/utils/oscap_docker_python/oscap_docker_util.py", line 151, in _scan
    return self.oscap_chroot("foo", "bar", chroot, *tmp_tuple)
  File "/home/jcerny/openscap/utils/oscap_docker_python/oscap_docker_util.py", line 130, in oscap_chroot
    self._cleanup_by_path(chroot_path)
  File "/home/jcerny/openscap/utils/oscap_docker_python/oscap_docker_util.py", line 177, in _cleanup_by_path
    DM._clean_temp_container_by_path(_no_rootfs)
  File "/usr/lib/python3.6/site-packages/Atomic/mount.py", line 742, in _clean_temp_container_by_path
    self.d.remove_container(short_cid)
  File "/usr/lib/python3.6/site-packages/Atomic/client.py", line 66, in newfunc
    result = attr(*args, **kwargs)
  File "/usr/lib/python3.6/site-packages/docker/utils/decorators.py", line 19, in wrapped
    return f(self, resource_id, *args, **kwargs)
  File "/usr/lib/python3.6/site-packages/docker/api/container.py", line 982, in remove_container
    self._raise_for_status(res)
  File "/usr/lib/python3.6/site-packages/docker/api/client.py", line 222, in _raise_for_status
    raise create_api_error_from_http_exception(e)
  File "/usr/lib/python3.6/site-packages/docker/errors.py", line 31, in create_api_error_from_http_exception
    raise cls(e, response=response, explanation=explanation)
docker.errors.NotFound: 404 Client Error: Not Found ("No such container: tmpnxkmw9tj")
```

This is already with applied fix from PR #795 .

**Comment by jan-cerny** [comment]
The fix isn't as easy as to change directory path in https://github.com/OpenSCAP/openscap/blob/maint-1.2/utils/oscap_docker_python/oscap_docker_util.py#L262, beacuse if I do that, I'll get this:
```
Title
	gpgcheck Enabled In Main Dnf Configuration
Rule
	ensure_gpgcheck_globally_activated
Result
	notapplicable

Title
	gpgcheck Enabled For All Dnf Package Repositories
Rule
	ensure_gpgcheck_never_disabled
Result
	notapplicable

Title
	Disable Prelinking
Rule
	disable_prelink
Result
	notapplicable

Title
	Build and Test AIDE Database
Rule
	aide_build_database
Result
	notapplicable

Title
	Verify and Correct File Permissions with RPM
Rule
	rpm_verify_permissions
Result
	notapplicable

Title
	Verify File Hashes with RPM
Rule
	rpm_verify_hashes
Result
	notapplicable

Title
	Disable Kernel Support for USB via Bootloader Configuration
Rule
	bootloader_nousb_argument
Result
	notapplicable

Title
	Verify that Shared Library Files Have Restrictive Permissions
Rule
	file_permissions_library_dirs
Result
	notapplicable

Title
	Verify that Shared Library Files Have Root Ownership
Rule
	file_ownership_library_dirs
Result
	notapplicable

Title
	Verify that System Executables Have Restrictive Permissions
Rule
	file_permissions_binary_dirs
Result
	notapplicable

Title
	Verify that System Executables Have Root Ownership
Rule
	file_ownership_binary_dirs
Result
	notapplicable

Title
	Direct root Logins Not Allowed
Rule
	no_direct_root_logins
Result
	notapplicable

Title
	Virtual Console Root Logins Restricted
Rule
	securetty_root_login_console_only
Result
	notapplicable

Title
	Serial Port Root Logins Restricted
Rule
	restrict_serial_port_logins
Result
	notapplicable

Title
	Only Root Has UID 0
Rule
	accounts_no_uid_except_zero
Result
	notapplicable

Title
	Log In to Accounts With Empty Password Impossible
Rule
	no_empty_passwords
Result
	notapplicable

Title
	Password Hashes For Each Account Shadowed
Rule
	accounts_password_all_shadowed
Result
	notapplicable

Title
	All GIDs referenced in /etc/passwd Defined in /etc/group
Rule
	gid_passwd_group_same
Result
	notapplicable

Title
	netrc Files Do Not Exist
Rule
	no_netrc_files
Result
	notapplicable

Title
	Password Minimum Length
Rule
	accounts_password_minlen_login_defs
Result
	notapplicable

Title
	Password Minimum Age
Rule
	accounts_minimum_age_login_defs
Result
	notapplicable

Title
	Password Maximum Age
Rule
	accounts_maximum_age_login_defs
Result
	notapplicable

Title
	Password Warning Age
Rule
	accounts_password_warn_age_login_defs
Result
	notapplicable

Title
	Ensure All Accounts on the System Have Unique Names
Rule
	account_unique_name
Result
	notapplicable

Title
	Set Last Logon/Access Notification
Rule
	display_login_attempts
Result
	notapplicable

Title
	Ensure that Root's Path Does Not Include World or Group-Writable Directories
Rule
	accounts_root_path_dirs_no_write
Result
	notapplicable

Title
	Verify firewalld Enabled
Rule
	service_firewalld_enabled
Result
	notapplicable

Title
	Set Default firewalld Zone for Incoming Packets
Rule
	set_firewalld_default_zone
Result
	notapplicable

Title
	Enable Auditing for Processes Which Start Prior to the Audit Daemon
Rule
	bootloader_audit_argument
Result
	notapplicable

Title
	Configure auditd Number of Logs Retained
Rule
	auditd_data_retention_num_logs
Result
	notapplicable

Title
	Configure auditd Max Log File Size
Rule
	auditd_data_retention_max_log_file
Result
	notapplicable

Title
	Configure auditd max_log_file_action Upon Reaching Maximum Log Size
Rule
	auditd_data_retention_max_log_file_action
Result
	notapplicable

Title
	Configure auditd space_left Action on Low Disk Space
Rule
	auditd_data_retention_space_left_action
Result
	notapplicable

Title
	Configure auditd admin_space_left Action on Low Disk Space
Rule
	auditd_data_retention_admin_space_left_action
Result
	notapplicable

Title
	Configure auditd mail_acct Action on Low Disk Space
Rule
	auditd_data_retention_action_mail_acct
Result
	notapplicable

Title
	Configure auditd to use audispd's syslog plugin
Rule
	auditd_audispd_syslog_plugin_activated
Result
	notapplicable

Title
	Record attempts to alter time through adjtimex
Rule
	audit_rules_time_adjtimex
Result
	notapplicable

Title
	Record attempts to alter time through settimeofday
Rule
	audit_rules_time_settimeofday
Result
	notapplicable

Title
	Record Attempts to Alter Time Through stime
Rule
	audit_rules_time_stime
Result
	notapplicable

Title
	Record Attempts to Alter Time Through clock_settime
Rule
	audit_rules_time_clock_settime
Result
	notapplicable

Title
	Record Attempts to Alter the localtime File
Rule
	audit_rules_time_watch_localtime
Result
	notapplicable

Title
	Record Events that Modify User/Group Information
Rule
	audit_rules_usergroup_modification
Result
	notapplicable

Title
	Record Events that Modify the System's Network Environment
Rule
	audit_rules_networkconfig_modification
Result
	notapplicable

Title
	System Audit Logs Must Be Owned By Root
Rule
	file_ownership_var_log_audit
Result
	notapplicable

Title
	Record Events that Modify the System's Mandatory Access Controls
Rule
	audit_rules_mac_modification
Result
	notapplicable

Title
	Record Events that Modify the System's Discretionary Access Controls - chmod
Rule
	audit_rules_dac_modification_chmod
Result
	notapplicable

Title
	Record Events that Modify the System's Discretionary Access Controls - chown
Rule
	audit_rules_dac_modification_chown
Result
	notapplicable

Title
	Record Events that Modify the System's Discretionary Access Controls - fchmod
Rule
	audit_rules_dac_modification_fchmod
Result
	notapplicable

Title
	Record Events that Modify the System's Discretionary Access Controls - fchmodat
Rule
	audit_rules_dac_modification_fchmodat
Result
	notapplicable

Title
	Record Events that Modify the System's Discretionary Access Controls - fchown
Rule
	audit_rules_dac_modification_fchown
Result
	notapplicable

Title
	Record Events that Modify the System's Discretionary Access Controls - fchownat
Rule
	audit_rules_dac_modification_fchownat
Result
	notapplicable

Title
	Record Events that Modify the System's Discretionary Access Controls - fremovexattr
Rule
	audit_rules_dac_modification_fremovexattr
Result
	notapplicable

Title
	Record Events that Modify the System's Discretionary Access Controls - fsetxattr
Rule
	audit_rules_dac_modification_fsetxattr
Result
	notapplicable

Title
	Record Events that Modify the System's Discretionary Access Controls - lchown
Rule
	audit_rules_dac_modification_lchown
Result
	notapplicable

Title
	Record Events that Modify the System's Discretionary Access Controls - lremovexattr
Rule
	audit_rules_dac_modification_lremovexattr
Result
	notapplicable

Title
	Record Events that Modify the System's Discretionary Access Controls - lsetxattr
Rule
	audit_rules_dac_modification_lsetxattr
Result
	notapplicable

Title
	Record Events that Modify the System's Discretionary Access Controls - removexattr
Rule
	audit_rules_dac_modification_removexattr
Result
	notapplicable

Title
	Record Events that Modify the System's Discretionary Access Controls - setxattr
Rule
	audit_rules_dac_modification_setxattr
Result
	notapplicable

Title
	Record Attempts to Alter Logon and Logout Events
Rule
	audit_rules_login_events
Result
	notapplicable

Title
	Record Attempts to Alter Process and Session Initiation Information
Rule
	audit_rules_session_events
Result
	notapplicable

Title
	Ensure auditd Collects Unauthorized Access Attempts to Files (unsuccessful)
Rule
	audit_rules_unsuccessful_file_modification
Result
	notapplicable

Title
	Ensure auditd Collects Information on the Use of Privileged Commands
Rule
	audit_rules_privileged_commands
Result
	notapplicable

Title
	Ensure auditd Collects Information on Exporting to Media (successful)
Rule
	audit_rules_media_export
Result
	notapplicable

Title
	Ensure auditd Collects File Deletion Events by User
Rule
	audit_rules_file_deletion_events
Result
	notapplicable

Title
	Ensure auditd Collects System Administrator Actions
Rule
	audit_rules_sysadmin_actions
Result
	notapplicable

Title
	Ensure auditd Collects Information on Kernel Module Loading and Unloading
Rule
	audit_rules_kernel_module_loading
Result
	notapplicable

Title
	Make the auditd Configuration Immutable
Rule
	audit_rules_immutable
Result
	notapplicable

Title
	SSH Root Login Disabled
Rule
	sshd_disable_root_login
Result
	notapplicable

Title
	SSH Access via Empty Passwords Disabled
Rule
	sshd_disable_empty_passwords
Result
	notapplicable

Title
	SSH Idle Timeout Interval Used
Rule
	sshd_set_idle_timeout
Result
	notapplicable

Title
	SSH Client Alive Count Used
Rule
	sshd_set_keepalive
Result
	notapplicable

Title
	Enable the Chrony Daemon
Rule
	service_chronyd_enabled
Result
	notapplicable

Title
	Specify a Remote NTP Server
Rule
	chronyd_specify_remote_server
Result
	notapplicable

Title
	Enable the Audit Daemon
Rule
	service_auditd_enabled
Result
	notapplicable


Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/docker/api/client.py", line 220, in _raise_for_status
    response.raise_for_status()
  File "/usr/lib/python3.6/site-packages/requests/models.py", line 909, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 404 Client Error: Not Found for url: http+docker://localunixsocket/v1.26/containers/tmprbly4_x4?v=False&link=False&force=False

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./oscap-docker", line 115, in <module>
    args.func()
  File "./oscap-docker", line 51, in scan
    result = OS.scan(self.args.scan_target, self.unknown_args)
  File "/home/jcerny/openscap/utils/oscap_docker_python/oscap_docker_util.py", line 268, in scan
    self.helper._cleanup_by_path(_tmp_mnt_dir)
  File "/home/jcerny/openscap/utils/oscap_docker_python/oscap_docker_util.py", line 177, in _cleanup_by_path
    DM._clean_temp_container_by_path(_no_rootfs)
  File "/usr/lib/python3.6/site-packages/Atomic/mount.py", line 742, in _clean_temp_container_by_path
    self.d.remove_container(short_cid)
  File "/usr/lib/python3.6/site-packages/Atomic/client.py", line 66, in newfunc
    result = attr(*args, **kwargs)
  File "/usr/lib/python3.6/site-packages/docker/utils/decorators.py", line 19, in wrapped
    return f(self, resource_id, *args, **kwargs)
  File "/usr/lib/python3.6/site-packages/docker/api/container.py", line 982, in remove_container
    self._raise_for_status(res)
  File "/usr/lib/python3.6/site-packages/docker/api/client.py", line 222, in _raise_for_status
    raise create_api_error_from_http_exception(e)
  File "/usr/lib/python3.6/site-packages/docker/errors.py", line 31, in create_api_error_from_http_exception
    raise cls(e, response=response, explanation=explanation)
docker.errors.NotFound: 404 Client Error: Not Found ("No such container: tmprbly4_x4")

```

**Comment by jan-cerny** [comment]
I think this need a fix in Atomic as well. I think we need to have https://github.com/projectatomic/atomic/blob/master/Atomic/mount.py#L511, which is present only in method `DockerMount._mount_devicemapper()` copy to `DockerMount._mount_overlay()`

**Comment by jan-cerny** [comment]
https://github.com/projectatomic/atomic/pull/1063

**Comment by jan-cerny** [comment]
What are your opinions on this topic?

**Comment by yuumasato** [comment]
Not totally fixed by #799.

**Comment by jan-cerny** [comment]
But rest of the fix is up to Atomic, which is a different repository.
